### ISSUE TITLE
[3.1.0] Coffee cup does not fill with coffee after placing in CoffeeMaker with `putNearXY` + turning on

### ISSUE THREAD

**Issue by thomason-jesse** [allenai/ai2thor#747]
Verified in the following scenes, with attached screenshots. I can get more examples if needed.

`FloorPlan14_physics`
`FloorPlan23_physics`
`FloorPlan26_physics`
`FloorPlan27_physics`

![Screen Shot 2021-05-09 at 3 45 12 PM](https://user-images.githubusercontent.com/842508/117589506-c0defe80-b0de-11eb-9829-fe91da109d16.png)
![Screen Shot 2021-05-09 at 3 46 17 PM](https://user-images.githubusercontent.com/842508/117589507-c2102b80-b0de-11eb-9453-8b60636d14f9.png)
![Screen Shot 2021-05-09 at 3 52 37 PM](https://user-images.githubusercontent.com/842508/117589508-c2a8c200-b0de-11eb-81f0-813622cfc6d3.png)
![Screen Shot 2021-05-09 at 3 43 04 PM](https://user-images.githubusercontent.com/842508/117589509-c3415880-b0de-11eb-80b4-7b41506eb3af.png)

Anecdotally, I experienced this behavior previously when approximating the `putNearXY` flag via a combination of `DropHandObject`, `GetSpawnCoordinatesAboveReceptacle`, and `PlaceObjectAtPoint`. When the mug doesn't land at the ~exact right place on the CoffeeMaker, the automatic filling doesn't fire. If that's what's happening now, could we either

ideally 1) change the auto-filling / cooking / etc. behaviors to fire if the target object is on any part of the receptacle or

alternatively 2) change `putNearXY` to be ignored when dealing with appliances with really just one "valid" location like the CoffeeMaker or stove burners.

Thanks in advance for any thoughts!

**Comment by thomason-jesse** [comment]
Hi Winson, I've attached a script example that creates this behavior that should let you re-create it. The final line is a print showing that the simulator "believes" the mug has been filled, and an image from the view frame showing that it hasn't (at least in terms of animation).

This script is certainly not the "minimal" way to reproduce the behavior, but these are all the steps taken in my examples before I observe it.

![final_frame](https://user-images.githubusercontent.com/842508/118045673-13adf580-b32d-11eb-9477-0dddbbcf0e53.jpeg)

Script:

```

from ai2thor.controller import Controller
from PIL import Image

init_params = {'scene': 'FloorPlan26_physics', 'gridSize': 0.25, 'snapToGrid': True, 'visibilityDistance': 1.5, 'width': 900, 'height': 900, 'agentCount': 2, 'commit_id': '2cca88de394ff1fa3f85f9ee3585444552059da9'}
controller = Controller(**init_params)

ac = {'action': 'LookDown', 'agentId': 0, 'degrees': 30}
controller.step(**ac)
ac = {'action': 'LookDown', 'agentId': 1, 'degrees': 30}
controller.step(**ac)

ac = {'action': 'ToggleMapView'}
controller.step(**ac)
ac = {'action': 'ToggleMapView'}
controller.step(**ac)

ac = {'action': 'SetObjectPoses', 'objectPoses': [{'objectName': 'PepperShaker_573b73c5', 'rotation': {'x': 6.017724808771163e-05, 'y': 90.00078582763672, 'z': 5.13686791236978e-06}, 'position': {'x': -0.504004180431366, 'y': 0.8632610440254211, 'z': 3.050718069076538}}, {'objectName': 'SaltShaker_0540bb5c', 'rotation': {'x': -0.0, 'y': 90.0, 'z': 0.0}, 'position': {'x': -2.6447548866271973, 'y': 0.2922644019126892, 'z': 4.501236915588379}}, {'objectName': 'Cup_31709fb7', 'rotation': {'x': -0.0, 'y': 0.0, 'z': 0.0}, 'position': {'x': -2.94370174407959, 'y': 0.9309447407722473, 'z': 3.209641456604004}}, {'objectName': 'Potato_c0d351ab', 'rotation': {'x': -0.0, 'y': 90.0, 'z': 0.0}, 'position': {'x': -2.722590684890747, 'y': 0.6575518846511841, 'z': 4.352762699127197}}, {'objectName': 'Spatula_a8a34a93', 'rotation': {'x': -3.6034055028721923e-06, 'y': 180.0020751953125, 'z': 4.909133622277295e-06}, 'position': {'x': -0.40305158495903015, 'y': 1.0240614414215088, 'z': 0.20372259616851807}}, {'objectName': 'Lettuce_fabe0296', 'rotation': {'x': -0.0, 'y': 0.0, 'z': 0.0}, 'position': {'x': -2.1363205909729004, 'y': 1.836144208908081, 'z': 0.2058212012052536}}, {'objectName': 'Apple_bb0e1437', 'rotation': {'x': 1.403341910044412e-14, 'y': 0.0, 'z': 0.0}, 'position': {'x': -1.3533079624176025, 'y': 0.8498344421386719, 'z': 0.28182023763656616}}, {'objectName': 'Fork_7a68ae9b', 'rotation': {'x': 5.136873369337991e-06, 'y': 0.0007854717550799251, 'z': -6.017725172569044e-05}, 'position': {'x': -0.30398738384246826, 'y': 0.8690405488014221, 'z': 3.300724506378174}}, {'objectName': 'Apple_bb0e1437', 'rotation': {'x': -0.0, 'y': 90.0, 'z': 0.0}, 'position': {'x': -2.7438430786132812, 'y': 1.2483534812927246, 'z': 4.352762699127197}}, {'objectName': 'PepperShaker_573b73c5', 'rotation': {'x': -0.0, 'y': 90.0, 'z': 0.0}, 'position': {'x': -2.6810081005096436, 'y': 0.6257460713386536, 'z': 4.427000045776367}}, {'objectName': 'Cup_31709fb7', 'rotation': {'x': -0.0, 'y': 90.0, 'z': 0.0}, 'position': {'x': -2.8785648345947266, 'y': 0.29453468322753906, 'z': 4.649711608886719}}, {'objectName': 'Fork_7a68ae9b', 'rotation': {'x': 5.136873369337991e-06, 'y': 0.0007854717550799251, 'z': -6.017725172569044e-05}, 'position': {'x': -0.6040069460868835, 'y': 0.8690409064292908, 'z': 3.4257373809814453}}, {'objectName': 'Spatula_a8a34a93', 'rotation': {'x': -0.0, 'y': 180.0, 'z': 0.0}, 'position': {'x': -1.9860188961029053, 'y': 0.6650130748748779, 'z': 0.5192608833312988}}, {'objectName': 'Egg_7407abdb', 'rotation': {'x': 359.9833679199219, 'y': 0.00029122462728992105, 'z': 0.003302386961877346}, 'position': {'x': -2.7468483448028564, 'y': 0.32293039560317993, 'z': 4.350236415863037}}, {'objectName': 'Plate_5ac26ac3', 'rotation': {'x': -0.0, 'y': 90.0, 'z': -0.0}, 'position': {'x': -2.805756092071533, 'y': 1.1818571090698242, 'z': 4.352762699127197}}, {'objectName': 'Toaster_70e5cd67', 'rotation': {'x': -2.2515355340146925e-06, 'y': 269.9964599609375, 'z': -5.8498437283560634e-05}, 'position': {'x': -0.2860190272331238, 'y': 0.8615368604660034, 'z': 3.5210022926330566}}, {'objectName': 'GarbageCan_5753b359', 'rotation': {'x': 358.5364074707031, 'y': 26.77322006225586, 'z': 2.444348096847534}, 'position': {'x': -2.8156261444091797, 'y': -0.014672160148620605, 'z': 0.37150314450263977}}, {'objectName': 'Fork_7a68ae9b', 'rotation': {'x': 5.136873369337991e-06, 'y': 0.0007854717550799251, 'z': -6.017725172569044e-05}, 'position': {'x': -0.7040208578109741, 'y': 0.8690410852432251, 'z': 2.925703287124634}}, {'objectName': 'Bread_420dedde', 'rotation': {'x': 359.9587097167969, 'y': 359.9520263671875, 'z': 0.5705959796905518}, 'position': {'x': -2.8713619709014893, 'y': 0.7695293426513672, 'z': 2.407951831817627}}, {'objectName': 'Apple_bb0e1437', 'rotation': {'x': 3.6034055028721923e-06, 'y': 0.002075874013826251, 'z': -4.909133167529944e-06}, 'position': {'x': -0.40305158495903015, 'y': 1.0784653425216675, 'z': 0.13999560475349426}}, {'objectName': 'Lettuce_fabe0296', 'rotation': {'x': 0.0, 'y': 180.0, 'z': 0.0}, 'position': {'x': -2.8889217376708984, 'y': 0.6951488852500916, 'z': 4.501237392425537}}, {'objectName': 'Spatula_a8a34a93', 'rotation': {'x': -0.0, 'y': 90.0, 'z': 1.403341910044412e-14}, 'position': {'x': -1.266960620880127, 'y': 0.7985031008720398, 'z': 0.22654026746749878}}, {'objectName': 'Knife_3e088444', 'rotation': {'x': 359.4722595214844, 'y': 89.99915313720703, 'z': 359.954833984375}, 'position': {'x': -2.5046331882476807, 'y': 0.7561604976654053, 'z': 2.8222827911376953}}, {'objectName': 'Potato_c0d351ab', 'rotation': {'x': -0.0, 'y': 90.0, 'z': -0.0}, 'position': {'x': -2.8318028450012207, 'y': 0.3240702152252197, 'z': 4.3527631759643555}}, {'objectName': 'Tomato_156a2494', 'rotation': {'x': -0.0, 'y': 0.0, 'z': 0.0}, 'position': {'x': -0.21831925213336945, 'y': 0.7253376841545105, 'z': 0.4645734429359436}}, {'objectName': 'Cup_31709fb7', 'rotation': {'x': -0.0, 'y': 0.0, 'z': 0.0}, 'position': {'x': -1.7709681987762451, 'y': 0.9236510992050171, 'z': 0.4335513412952423}}, {'objectName': 'SaltShaker_0540bb5c', 'rotation': {'x': -1.1613701265868887e-13, 'y': 179.99966430664062, 'z': -3.9362745951621037e-08}, 'position': {'x': -0.5358996987342834, 'y': 0.9364967346191406, 'z': 4.2538862228393555}}, {'objectName': 'PepperShaker_573b73c5', 'rotation': {'x': -0.0, 'y': 90.0, 'z': -0.0}, 'position': {'x': -2.7641735076904297, 'y': 0.6257460713386536, 'z': 4.575474739074707}}, {'objectName': 'ButterKnife_e8fcbb64', 'rotation': {'x': 0.0025619312655180693, 'y': 8.056790102273226e-05, 'z': 0.0003626760153565556}, 'position': {'x': -2.210432291030884, 'y': 0.918315052986145, 'z': 0.6198135614395142}}, {'objectName': 'Microwave_d485f342', 'rotation': {'x': 3.6034055028721923e-06, 'y': 0.002075874013826251, 'z': -4.909133167529944e-06}, 'position': {'x': -0.3589544892311096, 'y': 0.916986346244812, 'z': 0.22394701838493347}}, {'objectName': 'CoffeeMachine_95f5a8d0', 'rotation': {'x': 0.00760048720985651, 'y': 89.64036560058594, 'z': 0.010260531678795815}, 'position': {'x': -2.938218355178833, 'y': 0.9253959059715271, 'z': 3.605862855911255}}, {'objectName': 'SoapBottle_f176a976', 'rotation': {'x': 6.017724808771163e-05, 'y': 90.00078582763672, 'z': 5.13686791236978e-06}, 'position': {'x': -0.3318164646625519, 'y': 0.879283607006073, 'z': 2.8978748321533203}}, {'objectName': 'Pan_2f052047', 'rotation': {'x': 6.017724808771163e-05, 'y': 90.00078582763672, 'z': 5.13686791236978e-06}, 'position': {'x': -0.3039925694465637, 'y': 0.8669077157974243, 'z': 2.9256980419158936}}, {'objectName': 'DishSponge_e6b56422', 'rotation': {'x': 0.24819299578666687, 'y': 30.058238983154297, 'z': 2.2784557342529297}, 'position': {'x': -0.9010124802589417, 'y': 0.7889032959938049, 'z': 0.2600328326225281}}, {'objectName': 'Pot_40dbec4e', 'rotation': {'x': -0.0, 'y': 0.0, 'z': 0.0}, 'position': {'x': -0.5358999967575073, 'y': 0.9289288520812988, 'z': 4.274199962615967}}, {'objectName': 'Spoon_8ed80887', 'rotation': {'x': 359.70159912109375, 'y': -0.004966553300619125, 'z': 359.8957214355469}, 'position': {'x': -1.8015310764312744, 'y': 0.6518693566322327, 'z': 0.3378402292728424}}, {'objectName': 'Mug_f2b785ac', 'rotation': {'x': 359.954833984375, 'y': -0.0012426135363057256, 'z': 0.5277353525161743}, 'position': {'x': -2.504364252090454, 'y': 0.7272849082946777, 'z': 2.6073083877563477}}, {'objectName': 'DiningTable_5ff5c255', 'rotation': {'x': 0.5277352929115295, 'y': 269.999267578125, 'z': 0.045129887759685516}, 'position': {'x': -2.734349012374878, 'y': -0.0023845434188842773, 'z': 2.6068780422210693}}, {'objectName': 'DiningTable_7804dda2', 'rotation': {'x': -6.0177233535796404e-05, 'y': 270.0008544921875, 'z': -5.136985237186309e-06}, 'position': {'x': -0.48800328373908997, 'y': -0.0032619237899780273, 'z': 3.182717800140381}}, {'objectName': 'Bowl_65ad762e', 'rotation': {'x': 359.8799743652344, 'y': 0.01282525435090065, 'z': 0.2872719168663025}, 'position': {'x': -2.6369283199310303, 'y': 0.7219997644424438, 'z': 2.867621421813965}}], 'placeStationary': True}
controller.step(**ac)

ac = {'action': 'DirtyObject', 'objectId': 'Cup|-02.88|+00.29|+04.65', 'forceAction': True}
controller.step(**ac)
ac = {'action': 'ToggleObjectOn', 'objectId': 'StoveKnob|-00.06|+01.09|+04.78', 'forceAction': True}
controller.step(**ac)
ac = {'action': 'DirtyObject', 'objectId': 'Plate|-02.81|+01.18|+04.35', 'forceAction': True}
controller.step(**ac)
ac = {'action': 'ToggleObjectOn', 'objectId': 'Microwave|-00.36|+00.92|+00.22', 'forceAction': True}
controller.step(**ac)
ac = {'action': 'ToggleObjectOn', 'objectId': 'CoffeeMachine|-02.94|+00.93|+03.61', 'forceAction': True}
controller.step(**ac)
ac = {'action': 'DirtyObject', 'objectId': 'Pan|-00.30|+00.87|+02.93', 'forceAction': True}
controller.step(**ac)
ac = {'action': 'FillObjectWithLiquid', 'objectId': 'Bowl|-02.64|+00.72|+02.87', 'forceAction': True, 'fillLiquid': 'water'}
controller.step(**ac)
ac = {'action': 'DirtyObject', 'objectId': 'Bowl|-02.64|+00.72|+02.87', 'forceAction': True}
controller.step(**ac)

ac = {'action': 'Teleport', 'agentId': 0, 'x': -2.0, 'y': 0.9015909433364868, 'z': 2.5, 'rotation': {'x': -0.0, 'y': 270.0, 'z': 0.0}, 'horizon': 0.0}
controller.step(**ac)
ac = {'action': 'Teleport', 'agentId': 1, 'x': -1.0, 'y': 0.9015909433364868, 'z': 4.5, 'rotation': {'x': -0.0, 'y': 180.0, 'z': 0.0}, 'horizon': 30.00000762939453}
controller.step(**ac)

ac = {'action': 'MakeObjectsOfTypeUnbreakable', 'objectType': 'Mug'}
controller.step(**ac)
ac = {'action': 'MakeObjectsOfTypeUnbreakable', 'objectType': 'Window'}
controller.step(**ac)
ac = {'action': 'MakeObjectsOfTypeUnbreakable', 'objectType': 'Plate'}
controller.step(**ac)
ac = {'action': 'MakeObjectsOfTypeUnbreakable', 'objectType': 'Bowl'}
controller.step(**ac)
ac = {'action': 'MakeObjectsOfTypeUnbreakable', 'objectType': 'Cup'}
controller.step(**ac)
ac = {'action': 'MakeObjectsOfTypeUnbreakable', 'objectType': 'Egg'}
controller.step(**ac)

ac = {'action': 'LookDown', 'agentId': 0, 'degrees': 30, 'forceAction': True}
controller.step(**ac)
ac = {'action': 'Pass', 'agentId': 0, 'renderObjectImage': True}
controller.step(**ac)
ac = {'action': 'GetCoordinateFromRaycast', 'x': 0.5659722222222222, 'y': 0.7586805555555556, 'agentId': 0}
controller.step(**ac)
ac = {'action': 'PickupObject', 'agentId': 0, 'objectId': 'Mug|-02.50|+00.73|+02.61'}
controller.step(**ac)
ac = {'action': 'Pass', 'agentId': 0, 'renderObjectImage': True}
controller.step(**ac)
ac = {'action': 'Pass', 'agentId': 0, 'renderObjectImage': True}
controller.step(**ac)
ac = {'action': 'GetCoordinateFromRaycast', 'x': 0.9392361111111112, 'y': 0.5138888888888888, 'agentId': 0}
controller.step(**ac)
ac = {'action': 'PutObject', 'agentId': 0, 'forceAction': True, 'placeStationary': True, 'x': 0.9392361111111112, 'y': 0.5138888888888888, 'putNearXY': True}
controller.step(**ac)
ac = {'action': 'RotateRight', 'agentId': 0, 'degrees': 90, 'forceAction': True}
controller.step(**ac)
ac = {'action': 'Pass', 'agentId': 0, 'renderObjectImage': True}
controller.step(**ac)
ac = {'action': 'GetCoordinateFromRaycast', 'x': 0.12152777777777778, 'y': 0.4184027777777778, 'agentId': 0}
controller.step(**ac)
ac = {'action': 'ToggleObjectOff', 'agentId': 0, 'objectId': 'CoffeeMachine|-02.94|+00.93|+03.61'}
controller.step(**ac)
ac = {'action': 'Pass', 'agentId': 0, 'renderObjectImage': True}
controller.step(**ac)
ac = {'action': 'GetCoordinateFromRaycast', 'x': 0.11631944444444445, 'y': 0.4184027777777778, 'agentId': 0}
controller.step(**ac)
ac = {'action': 'ToggleObjectOn', 'agentId': 0, 'objectId': 'CoffeeMachine|-02.94|+00.93|+03.61'}
controller.step(**ac)
ac = {'action': 'Pass', 'agentId': 0, 'renderObjectImage': True}
controller.step(**ac)
ac = {'action': 'GetCoordinateFromRaycast', 'x': 0.2222222222222222, 'y': 0.4392361111111111, 'agentId': 0}
controller.step(**ac)
ac = {'action': 'PickupObject', 'agentId': 0, 'objectId': 'Mug|-02.50|+00.73|+02.61'}
e = controller.step(**ac)

print('mug filled with liquid',
      [obj['isFilledWithLiquid'] for obj in e.metadata['objects'] if obj['objectId'] == 'Mug|-02.50|+00.73|+02.61'][0])

f = controller.last_event.events[0].frame
pil_img = Image.fromarray(f)
pil_img.save('final_frame.jpeg', format="jpeg")

```

**Comment by winthos** [comment]
Ok, quick update on this issue.

I've been able to replicate your exact action sequence in-editor, but even so I'm still unable to replicate the results you are seeing. Here is a sample image from the editor after executing your example

![Coffee](https://user-images.githubusercontent.com/17502636/118564559-5ef94700-b725-11eb-8369-f88feda257d6.PNG)

I've stepped through all the logic both with and without using `putNearXY` and I can't seem to find any spot where this mismatch in expected state is happening. Basically, first a target receptacle is selected either via `objectId` or an `XY` screen coordinate. Then, after that target receptacle is locked in, some object is placed in/on that receptacle. The thing with the Coffee Maker is, it is an object specific receptacle that will automatically fail if any other object is attempted to be placed in it via `PutObject`. So, in this case, where you are placing a mug via `PutObject`, the value of `putNearXY` is actually not used at all because there is a check specifically for "object specific receptacles" that doesn't use this bool at all. This makes sense, as placing a mug into a coffee maker only has a single valid point where it could be attached, so there is no "nearXY" at all because there is only a single valid attachment point.

As a sanity check, i've recruited other team members to also try and run your action sequence to see what results they get, but at least so far I'm thoroughly stumped as to what is happening since I can't replicate your state at all.

I think once we solve this mystery, a larger refactor of allowing a coffee machine to be able to fill a mug not placed explicitly via `PutObject` is a change that should be made, so that is still the plan. But first, we need to solve this initial mismatch so that any bugs in the current behavior are fixed.
